activejob/lib/active_job/queue_adapters/async_adapter.rb

```diff
@@ -1,22 +1,118 @@
+      def initialize(**executor_options)
+        @scheduler = Scheduler.new(**executor_options)
+      end
+
       def enqueue(job) #:nodoc:
-        ActiveJob::AsyncJob.enqueue(job.serialize, queue: job.queue_name)
+        @scheduler.enqueue JobWrapper.new(job), queue_name: job.queue_name
       end
 
       def enqueue_at(job, timestamp) #:nodoc:
-        ActiveJob::AsyncJob.enqueue_at(job.serialize, timestamp, queue: job.queue_name)
+        @scheduler.enqueue_at JobWrapper.new(job), timestamp, queue_name: job.queue_name
+      end
+
+      # Gracefully stop processing jobs. Finishes in-progress work and handles
+      # any new jobs following the executor's fallback policy (`caller_runs`).
+      # Waits for termination by default. Pass `wait: false` to continue.
+      def shutdown(wait: true)
+        @scheduler.shutdown wait: wait
+      end
+
+      def immediate=(immediate)
+        @scheduler.immediate = immediate
+      end
+
+      # Note that we don't actually need to serialize the jobs since we're
+      # performing them in-process, but we do so anyway for parity with other
+      # adapters and deployment environments. Otherwise, serialization bugs
+      # may creep in undetected.
+      class JobWrapper #:nodoc:
+        def initialize(job)
+          job.provider_job_id = SecureRandom.uuid
+          @job_data = job.serialize
+        end
+
+        def perform
+          Base.execute @job_data
+        end
+      end
+
+      # Keep the AJ adapter and its backend separate rather than baking this
+      # implemention in directly. This is a support class, not for public use.
+      class Scheduler #:nodoc:
+        DEFAULT_EXECUTOR_OPTIONS = {
+          min_threads:     0,
+          max_threads:     Concurrent.processor_count,
+          auto_terminate:  true,
+          idletime:        60, # 1 minute
+          max_queue:       0, # unlimited
+          fallback_policy: :caller_runs # shouldn't matter -- 0 max queue
+        }.freeze
+
+        attr_accessor :immediate
+
+        def initialize(**options)
+          self.immediate = false
+          @immediate_executor = Concurrent::ImmediateExecutor.new
+          @executor = Concurrent::ThreadPoolExecutor.new(DEFAULT_EXECUTOR_OPTIONS.merge(options))
+        end
+
+        def enqueue(job, queue_name:) #:nodoc:
+          executor_for(queue_name).post(job, &:perform)
+        end
+
+        def enqueue_at(job, timestamp, queue_name:) #:nodoc:
+          delay = timestamp - Time.current.to_f
+          if delay > 0
+            Concurrent::ScheduledTask.execute(delay, args: [job], executor: executor_for(queue_name), &:perform)
+          else
+            enqueue(job, queue_name: queue_name)
+          end
+        end
+
+        def shutdown(wait: true)
+          @executor.shutdown
+          @executor.wait_for_termination if wait
+        end
+
+        def executor_for(queue_name)
+          executor
+        end
```

Comment:
The old backend had separate executors per queue. The indirection here communicates intent that the executor may vary by queue and acts as a mental speed bump against refactoring that'd rule out executor selection.